### PR TITLE
Proposal: secret, configmap, downwardAPI all in one volume

### PR DESCRIPTION
**Pull request by jpeeler**
This is a proposal for adding a way of projecting secrets, configmaps, and downward API into the same volume.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35313)

### CODE REVIEW

File: docs/proposals/all-in-one-volume.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    system:
+      items:
+      - secretName: mysecret
+      - configMapName: myconfigmap
+```
+
+Here the assumption is that the data values configured for each resource is the
+same. The contents of /system-volume may look like this:
+
+/system-volume/1-very-generic
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: I don't think a numeric prefix is especially usable.  How about:

1.  `<type of resource>-<resource name>-<key>`
2.  `<resource name>-<key>`

File: docs/proposals/all-in-one-volume.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+### Additional required modifications
+
+Add to VolumeSource struct:
+
+```go
+SystemVolume *SystemVolumeSource `json:"system,omitempty"`
+```
+
+The appropriate conversion code would need to be generated for v1 as well.
+
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: What about validations?

Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: What about the new volume plugin code?

File: docs/proposals/all-in-one-volume.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+## Motivation
+
+Users often need to build directories that contain multiple types of
+configuration and secret data. For example, a configuration directory for some
+software package may contain both config files and credentials. Currently, there
+is no way to achieve this in Kubernetes without scripting inside of a container.
+
+## Constraints and Assumptions
+
+1.  The volume types must remain unchanged for backward compatability.
```
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: fixed

File: docs/proposals/all-in-one-volume.md (hunk 2 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: Added.

**Comment by tmrts** [comment]
@pmorie instead of adding a new "super-volume" type, isn't transforming the secrets/config-maps/etc. into objects that are stored in volumes?

**Comment by pmorie** [comment]
@tmrts 

> To elaborate, I'm proposing restructuring volume types where a volume would be mounted storage and a user would specify the contents of the volume (e.g. volume A contains secret X and configmap Y)

That is exactly what is being proposed here

**Comment by pmorie** [comment]
cc @thockin

File: docs/proposals/all-in-one-volume.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    system:
+      items:
+      - secretName: mysecret
+      - configMapName: myconfigmap
+```
+
+Here the assumption is that the data values configured for each resource is the
+same. The contents of /system-volume may look like this:
+
+/system-volume/secret-mysecret-very-generic
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: put this in a monospace block

File: docs/proposals/all-in-one-volume.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+data that is automatically projected, resolution will be done using the same
+numeric prefix addition as described in the first scenario.
+
+## Code changes
+
+### Proposed API objects
+
+```go
+type SystemVolumeSource struct {
+	Items []SystemVolumeProjection `json:"items,omitempty"`
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: This should have a 'DefaultMode' field

File: docs/proposals/all-in-one-volume.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+type SystemVolumeProjection struct {
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: indentation

File: docs/proposals/all-in-one-volume.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	DownwardAPI   string `json:"downwardAPI,omitempty"`
+}
+```
+
+### Additional required modifications
+
+Add to VolumeSource struct:
+
+```go
+SystemVolume *SystemVolumeSource `json:"system,omitempty"`
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: I usually put the enclosing struct with a comment saying 'other fields omitted'

File: docs/proposals/all-in-one-volume.md [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+data that is automatically projected, resolution will be done using the same
+numeric prefix addition as described in the first scenario.
+
+## Code changes
+
+### Proposed API objects
+
+```go
+type SystemVolumeSource struct {
+	Items       []SystemVolumeProjection `json:"items,omitempty"`
```
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: I don't understand why this is showing up like this. Both lines are using a tab.

Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: Spaces are a better idea than tabs in md docs in github

**Comment by bgrant0607** [comment]
ref #19762

**Comment by pmorie** [comment]
One thing that needs to be handled: secrets live on a tmpfs, configmap and downward API do not.  Part of the reason this is so is the currently, memory used by tmpfs for emptydir/secrets is charged to the kubelet's cgroup.

So, we need to ensure that secrets do not come to rest on the node in the new plugin.  For that to happen, we probably need to:

1. Figure out the right way to have a tmpfs and node-local storage live side-by-side in a single volume (currently, plugins that wrap emptyDir can have one or another)
2. Change `AtomicWriter` to support differentiating locations for pieces of the payload (so you can have the secret content live in a different place than configmap / downward API).

OR figure out how to charge the memory for the tmpfs against the right cgroup and just put it all on a tmpfs.

**Comment by derekwaynecarr** [comment]
I think any volume type that could mount a secret should always be backed by tmpfs, so in this context I would want configmap to be tmpfs backed.  So basically anything of this volume TPR would be charged to memory.

Per our chat, I think a conflict among secret and configmap keys that are not user remapped should result in a failure to start the pod.   I think 99% of the time it will be a result of user error, and doing any inferred conflict resolution will just hide errors.

Other than that, nice write-up!

**Comment by pmorie** [comment]
@kubernetes/api-review-team

**Comment by pmorie** [comment]
>  volume TPR 

I do not understand the meaning of 'TPR' in this context - that acronym typically means 'third party resource'.  Disambiguate?

**Comment by derekwaynecarr** [comment]
@pmorie - TPR was an unfortunate typo from my phone.

File: docs/proposals/all-in-one-volume.md [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+### Populating a single volume with pathing
+
+Currently it is possible to define the path within a volume for specific
+resources. Therefore the same is true for each resource contained within the
+new single volume.
+
+## Current State Overview
+
+The only way of utilizing secrets, configmaps, and downward API currently is
```
Comment (thockin) [inline docs/proposals/all-in-one-volume.md]: Not strictly true = subPath allows you to pull out individually named files and merge them.

File: docs/proposals/all-in-one-volume.md [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+### Proposed API objects
+
+```go
+type SystemVolumeSource struct {
+    Items       []SystemVolumeProjection `json:"items,omitempty"`
+    DefaultMode *int32                   `json:"defaultMode,omitempty"`
+}
+
+type SystemVolumeProjection struct {
+    SecretName    string `json:"secretName,omitempty"`
```
Comment (thockin) [inline docs/proposals/all-in-one-volume.md]: Why are these names, rather than the full *VolumeSource structs?

File: docs/proposals/all-in-one-volume.md [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+data that is automatically projected, resolution will be done using the same
+numeric prefix addition as described in the first scenario.
+
+## Code changes
+
+### Proposed API objects
+
+```go
+type SystemVolumeSource struct {
+    Items       []SystemVolumeProjection `json:"items,omitempty"`
```
Comment (thockin) [inline docs/proposals/all-in-one-volume.md]: Items should be reserved for individual files.  I'd call this "sources"

File: docs/proposals/all-in-one-volume.md [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+In the event that there is a collision due to a user specified path matching
+data that is automatically projected, resolution will be done using the same
+numeric prefix addition as described in the first scenario.
+
+## Code changes
+
+### Proposed API objects
+
+```go
+type SystemVolumeSource struct {
```
Comment (thockin) [inline docs/proposals/all-in-one-volume.md]: This name is bad.  It doesn't convey any information.

"Projections" or something like that would be better

**Review by thockin — COMMENTED** [review]
This doesn't handle merging with non-projected dirs.  what about #33179

File: docs/proposals/all-in-one-volume.md (hunk 10 above)
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: @jpeeler I had thought that we discussed using the existing volume sources - and could swear that that API was in an earlier version of this document.  Do you remember things the same way?

File: docs/proposals/all-in-one-volume.md (hunk 11 above)
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: 'sources' works for me

**Comment by pmorie** [comment]
@thockin 

> This doesn't handle merging with non-projected dirs. what about #33179

I have an idea for how to do this, but haven't had a chance to put it into a proposal yet.  The TLDR is, let's make it possible to put all the system projections into a single dir and then allow you to union mount that onto another part of a container's FS.

File: docs/proposals/all-in-one-volume.md (hunk 12 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: I used "SystemProjections" to stick with the system theme, but can shorten it if necessary.

File: docs/proposals/all-in-one-volume.md (hunk 11 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: Updated.

File: docs/proposals/all-in-one-volume.md (hunk 10 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: I failed to update this section as Paul and I had previously discussed. Fixed now.

File: docs/proposals/all-in-one-volume.md (hunk 9 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: I wasn't aware of this feature. I'll have to reword this to include this information.

**Comment by thockin** [comment]
NOt all installations use unionfs..  Are we making that a requirement?

File: docs/proposals/all-in-one-volume.md (hunk 9 above)
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: You're technically correct, @thockin, but you lose atomic update w/ subpath

**Comment by pmorie** [comment]
@thockin 

> NOt all installations use unionfs..  Are we making that a requirement?

I want to have NO dependency required on any fs :)

Keyword here is 'emulate' union mounts.

File: docs/proposals/all-in-one-volume.md [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+    image: busybox
+    volumeMounts:
+    - name: all-in-one
+      mountPath: "/system-volume"
+      readOnly: true
+  volumes:
+  - name: all-in-one
+    system:
+      sources:
+      - secretName: mysecret
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: this example needs to be updated to be consistent with the API you proposed

File: docs/proposals/all-in-one-volume.md [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: this does not have multiple secrets

File: docs/proposals/all-in-one-volume.md [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+    volumeMounts:
+    - name: all-in-one
+      mountPath: "/system-volume"
+      readOnly: true
+  volumes:
+  - name: all-in-one
+    system:
+      sources:
+      - secret:
+          secretName: mysecret
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: Seeing this, I think we might want to make this 'name' in the all-in-one volume.  We cannot change the existing field for backward compatibility reasons, but we can make this one right.

File: docs/proposals/all-in-one-volume.md [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+          items:
+            - key: config
+              path: my-group/data
+```
+
+### Collisions when one path is explicit and the other is automatically projected
+
+In the event that there is a collision due to a user specified path matching
+data that is automatically projected, resolution will be done using the same
+numeric prefix addition as described in the first scenario.
```
Comment (pmorie) [inline docs/proposals/all-in-one-volume.md]: Needs update

File: docs/proposals/all-in-one-volume.md (hunk 15 above)
Comment (thockin) [inline docs/proposals/all-in-one-volume.md]: you mean turn it into a LocalObjectRef, yeah, probably

File: docs/proposals/all-in-one-volume.md (hunk 16 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: Do you mean throw out all the resolution attempts and just make any conflict a validation error?

File: docs/proposals/all-in-one-volume.md (hunk 15 above)
Comment (jpeeler) [inline docs/proposals/all-in-one-volume.md]: added

**Comment by thockin** [comment]
If @pmorie "has ideas" about a merge-volume and so do I, do we want to proceed with this proposal?

**Comment by jpeeler** [comment]
@thockin, I believe the plan is to use this proposal as a stepping stone as suggested in  https://github.com/kubernetes/kubernetes/issues/33179#issuecomment-257421366. In any case, this work couldn't hurt future development for merging with non-projected directories, right?

**Comment by thockin** [comment]
My question is whether we want this volume driver AND a merge driver that
does all the same stuff plus-plus

**Comment by pmorie** [comment]
@thockin I view the union-mount piece as orthogonal or buildable upon this.  I think of the union-mount part as a property of the `VolumeMount` rather than a specific plugin, but I'm willing to hear arguments that I'm wrong.

**Comment by pmorie** [comment]
@bgrant0607 when I mentioned this at kubecon, I don't think you were convinced about it.  I'm wondering if we need to discuss this before Jeff moves further.

cc @thockin @jpeeler

**Comment by bgrant0607** [comment]
@pmorie @jpeeler 

I don't understand how this moves us closer to #33179, for which I've seen more use cases (e.g., /etc, /etc/foo, /home/foo, and #831 + Secret/ConfigMap/DownwardAPI).

sort of related: #30716

**Comment by thockin** [comment]
My point is that this is strictly a subset of merge volumes.  If we're
eventually going to handle that, this is redundant, no?

**Comment by pmorie** [comment]
@bgrant0607 @thockin My rationale for this is that it is probably more intuitive once we have a union-mount/merge mechanism to put all of the system-originated pieces of data for a single projection into the same volume.  Looking at the two different possibilities for this:

1.  User defines a single volume that contains the system-originated content they want for merge into the container's filesystem
2.  User defines potentially n volumes that contain the system-originated content they want for merge into the container's filesystem and also n volume-mounts that merge to the same place

I have to believe that (1) is the better user experience, and will make for pod specs that are far easier to read and write than (2).

TBH, knowing what I now know about where this functionality has settled, if we could do it over again, I would have planned an all-in-one volume from the start.

Any thoughts?

**Comment by thockin** [comment]
I don't think (1) is much less verbose than (2), a little but not much.
The question is whether it is overall a better UX for people to use this
mega-volume instead of the discrete per-source volumes.

If we can structure code so that the per-source volumes consume a pkg from
the mega volume, then maintenance is easy, and we can deprecate teh
per-source volumes in favor of this mega volume.

But it's all about the UX

**Comment by jpeeler** [comment]
@thockin 

I didn't realize that deprecation of existing plugins was a possible outcome of completing this work. Currently the WIP implementation I have calls into the other drivers to generate the payloads to hand off to the atomic writer. (Not to trivialize this work, but most of the rest of the driver is just filling out the VolumePlugin and Mounter interfaces.) If reversing those dependencies will allow this proposal to pass then I'll do it, no problem.

If deprecating the individual plugins is on the table, then maintenance could be even easier. But either way I think the system volume driver is nicer for the user in most cases (especially in what I think is the strongest case of easily using configmaps with secrets).

I think I need to rewrite some of the proposal regarding conflicts... but do we have consensus now that the new all in one driver is a good, well received goal?

**Comment by thockin** [comment]
I guess I can live with it.  I don't want multiple drivers forever, so I
would argue that we propose a deprecation for the 3 standalone drivers.
Deprecation can't happen until at least API v2, so it's not like it's going
to disappear.  But if we deprecate, we have the option to eventually remove
it if/when v2 happens.

**Comment by pmorie** [comment]
@thockin @jpeeler 

I would also like to move away from separate drivers that do basically the same thing.  +1 for all in one and deprecate existing drivers.

With that said, we should get some feedback from the community on the API to ensure people like it and it makes sense.  It's extremely easy to get tunnel vision on matters of usability as the person making the changes (of course it makes sense to you) and we have the opportunity now.

**Comment by erictune** [comment]
The proposal mentioned the example where you want to put config and secret values into the same single file (`~/.my.cnf`) but I don't see how this is possible for this proposal.  Am I missing something?

**Comment by pmorie** [comment]
@erictune That is not actually an aim of this proposal, so perhaps that is a typo

cc @jpeeler

**Comment by jpeeler** [comment]
Since this has been moved over to the new community repo, closing here.
